src/AspNetCoreController/Example/Startup.cs

```diff
@@ -23,18 +24,23 @@ public Startup(IConfiguration configuration)
 
         public void ConfigureServices(IServiceCollection services)
         {
-            services.AddGraphQL()
+            services.AddGraphQL(b => b
                 .AddSchema<StarWarsSchema>()
                 .AddSystemTextJson()
                 .AddValidationRule<InputValidationRule>()
                 .AddGraphTypes(typeof(StarWarsSchema).Assembly)
-                .AddMetrics(_ => true, (provider, _) => provider.GetRequiredService<IOptions<GraphQLSettings>>().Value.EnableMetrics);
+                .AddDocumentExecuter<ApolloTracingDocumentExecuter>()
+                .AddMetrics(_ => true, (provider, _) => provider.GetRequiredService<IOptions<GraphQLSettings>>().Value.EnableMetrics));
```

Comment:
Every time I see these 2 methods, then I doubt whether it was worth giving the opportunity to independently set up. I understand that `AddMetrics` adds some additional info into `ExecutionResult.Perf` object and `ApolloTracingDocumentExecuter` just copy this info (modifying slightly) into `ExecutionResult.Extensions` object in "tracing" bucket. @Shane32  what do you think?

Reply:
In short, I'm not satisfied either.

Ideally, I think `AddMetrics()` should be enough, and perhaps one overload that is conditional upon the service provider so it can be enabled or disabled depending on a runtime condition.  This is simple enough when the runtime condition is a compilation switch, such as checking if the environment is a development or checking environment variables, but if the condition is based on the `HttpContext.User` or some runtime condition, then this requires that middleware is initialized prior to the schema being initialized -- and the middleware requires significant processing overhead to all requests even if the metrics are not returned in the response.

Within the call to AddMetrics, I would like to see it add a document listener which will append the metrics to the result.  However, we may need to add another method to the document listener.  Or accept the fact that metrics may not include the execution time taken by other document listeners, and similar situations where the document listener method will not run.

Finally, I'm not happy that the metrics information is so tightly integrated to the ResolveFieldContext class.  I would rather see that within AddMetrics, the added middleware contains a reference to an AsyncLocal which is used to pull the current metrics data instance.

Reply:
One more note - one can use metrics internally without exposing it to client via extensions.

Reply:
True.  But at that point someone would be writing their own document listener or document executer to actually use the data, and if that's the case, they can write their own builder extension method to perform the necessary registrations.  Potentially we could have an overload to AddMetrics called `AddMetrics<TListener>` which would add the specified listener.

The builder methods are supposed to be easy-to-use helper methods to help configure the GraphQL.NET pipeline without needing to know all of the intricate details and need not cover every single configuration scenario.

Reply:
https://github.com/graphql-dotnet/graphql-dotnet/issues/3112